docs/building.md

```diff
@@ -0,0 +1,228 @@
+```
+
+Precompiled Abseil libraries are also available, so the above can also be done
+as:
+
+```makefile
+PROXY_WASM_CPP_SDK=/sdk
+CPP_API:=${PROXY_WASM_CPP_SDK}
+CPP_CONTEXT_LIB = ${CPP_API}/proxy_wasm_intrinsics.cc
+ABSL = /root/abseil-cpp
+ABSL_LIBS = ${ABSL}/absl/strings/libabsl_strings.a ${ABSL}/absl/strings/libabsl_strings_internal.a  ${ABSL}/absl/strings/libabsl_str_format_internal.a
+
+all: plugin.wasm
+
+%.wasm %.wat: %.cc ${CPP_API}/proxy_wasm_intrinsics.h ${CPP_API}/proxy_wasm_enums.h ${CPP_API}/proxy_wasm_externs.h ${CPP_API}/proxy_wasm_api.h ${CPP_API}/proxy_wasm_intrinsics.js ${CPP_CONTEXT_LIB}
+        ls /root
+                em++ --no-entry -s EXPORTED_FUNCTIONS=['_malloc'] --std=c++17 -O3 -flto -I${CPP_API} -I${CPP_API}/google/protobuf -I/usr/local/include -I${ABSL} --js-library ${CPP_API}/proxy_wasm_intrinsics.js  $*.cc ${CPP_API}/proxy_wasm_intrinsics.pb.cc ${CPP_CONTEXT_LIB} ${CPP_API}/libprotobuf.a ${ABSL_LIBS} -o $*.wasm
+```
+
+### Ownership of the resulting .wasm files
+
+The compiled files may be owned by root.  To chown them, add the follow lines to
+the Makefile and docker invocation:
+
+```makefile
+PROXY_WASM_CPP_SDK=/sdk
+
+all: myproject.wasm
+  chown ${uid}.${gid} $^
+
+include ${PROXY_WASM_CPP_SDK}/Makefile.base_lite
+```
+
+Invocation file (e.g. build.sh):
+
+```bash
+#!/bin/bash
+docker run -e uid="$(id -u)" -e gid="$(id -g)" -v $PWD:/work -w /work wasmsdk:v2 /build_wasm.sh
+```
+
+## Dependencies for building Wasm modules:
+
+If you do not wish to use the Docker image, the dependencies can be installed by
+script (sdk\_container.sh), or by hand.
+
+### protobuf v3.9.1
+
+You must install the version of protobuf on your build system that matches the
+libprotobuf.a files (without any patches) so that the generated code matches the
+.a library.  Currently this is based on tag v3.9.1 of
+https://github.com/protocolbuffers/protobuf.
+
+```bash
+git clone https://github.com/protocolbuffers/protobuf
+cd protobuf
+git checkout v3.9.1
+git submodule update --init --recursive
+./autogen.sh
+./configure
+make
+make check
+sudo make install
+```
+
+### Emscripten
+
+```bash
+git clone https://github.com/emscripten-core/emsdk.git
+cd emsdk
+./emsdk update-tags
+./emsdk install 3.1.7
+./emsdk activate 3.1.7
+
+source ./emsdk_env.sh
+```
+
+It is possible later versions will work, e.g.
+
+```bash
+./emsdk update-tags
+./emsdk install latest
+./emsdk activate latest
+```
+
+However 3.1.7 is known to work.
+
+### Rebuilding the libprotobuf.a files
+
+If want to rebuild the libprotobuf.a files or use a different version see the
+instructions at https://github.com/kwonoj/protobuf-wasm. Commit
+4bba8b2f38b5004f87489642b6ca4525ae72fe7f works for protobuf v3.9.x.
```

Comment:
Do we still need this hack and documentation for protobufs?

It looks that upstream added support for building with Emscripten: https://github.com/protocolbuffers/protobuf/commit/8e869f76275253726da5a712de4d4a1b4de22ca2

Reply:
Most likely not. I've kept the instructions in place for now, until #159 is fixed, but with an added note that the hack should only be needed for protobuf versions earlier than v3.15.